diff --git a/lib/api/ExtractorConfig.js b/lib/api/ExtractorConfig.js
index 31b46f8b51a2a93c2e538d67cd340e3c6897e242..d2cab369fc53cbd35bcb0c465783f851f1fd5f42 100644
--- a/lib/api/ExtractorConfig.js
+++ b/lib/api/ExtractorConfig.js
@@ -668,6 +668,6 @@ ExtractorConfig.FILENAME = 'api-extractor.json';
  */
 ExtractorConfig._tsdocBaseFilePath = path.resolve(__dirname, '../../extends/tsdoc-base.json');
 ExtractorConfig._defaultConfig = node_core_library_1.JsonFile.load(path.join(__dirname, '../schemas/api-extractor-defaults.json'));
-ExtractorConfig._declarationFileExtensionRegExp = /\.d\.ts$/i;
+ExtractorConfig._declarationFileExtensionRegExp = /\.d\.(m|c)?ts$/i;
 exports.ExtractorConfig = ExtractorConfig;
 //# sourceMappingURL=ExtractorConfig.js.map
\ No newline at end of file
