diff --git a/lib/api/ExtractorConfig.js b/lib/api/ExtractorConfig.js
index a37db0d564a5662df78055ded63069a4b2706bb1..158d4b121fa0c7cf15c3a52570218a2fe67fc567 100644
--- a/lib/api/ExtractorConfig.js
+++ b/lib/api/ExtractorConfig.js
@@ -669,5 +669,5 @@ ExtractorConfig.FILENAME = 'api-extractor.json';
  */
 ExtractorConfig._tsdocBaseFilePath = path.resolve(__dirname, '../../extends/tsdoc-base.json');
 ExtractorConfig._defaultConfig = node_core_library_1.JsonFile.load(path.join(__dirname, '../schemas/api-extractor-defaults.json'));
-ExtractorConfig._declarationFileExtensionRegExp = /\.d\.ts$/i;
+ExtractorConfig._declarationFileExtensionRegExp = /\.d\.(m|c)?ts$/i;
 //# sourceMappingURL=ExtractorConfig.js.map
\ No newline at end of file
