// @parcel/source-map/parcel_sourcemap_wasm/dist-web/parcel_sourcemap_wasm.js

let wasm;

const heap = new Array(32).fill(undefined);

heap.push(undefined, null, true, false);

function getObject(idx) { return heap[idx]; }

let WASM_VECTOR_LEN = 0;

let cachegetUint8Memory0 = null;
function getUint8Memory0() {
    if (cachegetUint8Memory0 === null || cachegetUint8Memory0.buffer !== wasm.memory.buffer) {
        cachegetUint8Memory0 = new Uint8Array(wasm.memory.buffer);
    }
    return cachegetUint8Memory0;
}

let cachedTextEncoder = new TextEncoder('utf-8');

const encodeString = (typeof cachedTextEncoder.encodeInto === 'function'
    ? function (arg, view) {
    return cachedTextEncoder.encodeInto(arg, view);
}
    : function (arg, view) {
    const buf = cachedTextEncoder.encode(arg);
    view.set(buf);
    return {
        read: arg.length,
        written: buf.length
    };
});

function passStringToWasm0(arg, malloc, realloc) {

    if (realloc === undefined) {
        const buf = cachedTextEncoder.encode(arg);
        const ptr = malloc(buf.length);
        getUint8Memory0().subarray(ptr, ptr + buf.length).set(buf);
        WASM_VECTOR_LEN = buf.length;
        return ptr;
    }

    let len = arg.length;
    let ptr = malloc(len);

    const mem = getUint8Memory0();

    let offset = 0;

    for (; offset < len; offset++) {
        const code = arg.charCodeAt(offset);
        if (code > 0x7F) break;
        mem[ptr + offset] = code;
    }

    if (offset !== len) {
        if (offset !== 0) {
            arg = arg.slice(offset);
        }
        ptr = realloc(ptr, len, len = offset + arg.length * 3);
        const view = getUint8Memory0().subarray(ptr + offset, ptr + len);
        const ret = encodeString(arg, view);

        offset += ret.written;
    }

    WASM_VECTOR_LEN = offset;
    return ptr;
}

let cachegetInt32Memory0 = null;
function getInt32Memory0() {
    if (cachegetInt32Memory0 === null || cachegetInt32Memory0.buffer !== wasm.memory.buffer) {
        cachegetInt32Memory0 = new Int32Array(wasm.memory.buffer);
    }
    return cachegetInt32Memory0;
}

let heap_next = heap.length;

function addHeapObject(obj) {
    if (heap_next === heap.length) heap.push(heap.length + 1);
    const idx = heap_next;
    heap_next = heap[idx];

    heap[idx] = obj;
    return idx;
}

function dropObject(idx) {
    if (idx < 36) return;
    heap[idx] = heap_next;
    heap_next = idx;
}

function takeObject(idx) {
    const ret = getObject(idx);
    dropObject(idx);
    return ret;
}

let cachedTextDecoder = new TextDecoder('utf-8', { ignoreBOM: true, fatal: true });

cachedTextDecoder.decode();

function getStringFromWasm0(ptr, len) {
    return cachedTextDecoder.decode(getUint8Memory0().subarray(ptr, ptr + len));
}

let cachegetUint32Memory0 = null;
function getUint32Memory0() {
    if (cachegetUint32Memory0 === null || cachegetUint32Memory0.buffer !== wasm.memory.buffer) {
        cachegetUint32Memory0 = new Uint32Array(wasm.memory.buffer);
    }
    return cachegetUint32Memory0;
}

function passArray32ToWasm0(arg, malloc) {
    const ptr = malloc(arg.length * 4);
    getUint32Memory0().set(arg, ptr / 4);
    WASM_VECTOR_LEN = arg.length;
    return ptr;
}

function _assertClass(instance, klass) {
    if (!(instance instanceof klass)) {
        throw new Error(`expected instance of ${klass.name}`);
    }
    return instance.ptr;
}
/**
*/
export class SourceMap {

    static __wrap(ptr) {
        const obj = Object.create(SourceMap.prototype);
        obj.ptr = ptr;

        return obj;
    }

    __destroy_into_raw() {
        const ptr = this.ptr;
        this.ptr = 0;

        return ptr;
    }

    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_sourcemap_free(ptr);
    }
    /**
    * @param {string} project_root
    * @param {any} buffer
    */
    constructor(project_root, buffer) {
        var ptr0 = passStringToWasm0(project_root, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        var ret = wasm.sourcemap_new(ptr0, len0, addHeapObject(buffer));
        return SourceMap.__wrap(ret);
    }
    /**
    * @returns {string}
    */
    getProjectRoot() {
        try {
            const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
            wasm.sourcemap_getProjectRoot(retptr, this.ptr);
            var r0 = getInt32Memory0()[retptr / 4 + 0];
            var r1 = getInt32Memory0()[retptr / 4 + 1];
            return getStringFromWasm0(r0, r1);
        } finally {
            wasm.__wbindgen_add_to_stack_pointer(16);
            wasm.__wbindgen_free(r0, r1);
        }
    }
    /**
    * @param {string} vlq_mappings
    * @param {any} sources
    * @param {any} sources_content
    * @param {any} names
    * @param {number} line_offset
    * @param {number} column_offset
    * @returns {any}
    */
    addVLQMap(vlq_mappings, sources, sources_content, names, line_offset, column_offset) {
        var ptr0 = passStringToWasm0(vlq_mappings, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        var ret = wasm.sourcemap_addVLQMap(this.ptr, ptr0, len0, addHeapObject(sources), addHeapObject(sources_content), addHeapObject(names), line_offset, column_offset);
        return takeObject(ret);
    }
    /**
    * @returns {any}
    */
    toVLQ() {
        var ret = wasm.sourcemap_toVLQ(this.ptr);
        return takeObject(ret);
    }
    /**
    * @returns {any}
    */
    getMappings() {
        var ret = wasm.sourcemap_getMappings(this.ptr);
        return takeObject(ret);
    }
    /**
    * @returns {any}
    */
    getSources() {
        var ret = wasm.sourcemap_getSources(this.ptr);
        return takeObject(ret);
    }
    /**
    * @returns {any}
    */
    getSourcesContent() {
        var ret = wasm.sourcemap_getSourcesContent(this.ptr);
        return takeObject(ret);
    }
    /**
    * @returns {any}
    */
    getNames() {
        var ret = wasm.sourcemap_getNames(this.ptr);
        return takeObject(ret);
    }
    /**
    * @param {string} name
    * @returns {number}
    */
    addName(name) {
        var ptr0 = passStringToWasm0(name, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        var ret = wasm.sourcemap_addName(this.ptr, ptr0, len0);
        return ret >>> 0;
    }
    /**
    * @param {string} source
    * @returns {number}
    */
    addSource(source) {
        var ptr0 = passStringToWasm0(source, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        var ret = wasm.sourcemap_addSource(this.ptr, ptr0, len0);
        return ret >>> 0;
    }
    /**
    * @param {number} index
    * @returns {string}
    */
    getName(index) {
        try {
            const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
            wasm.sourcemap_getName(retptr, this.ptr, index);
            var r0 = getInt32Memory0()[retptr / 4 + 0];
            var r1 = getInt32Memory0()[retptr / 4 + 1];
            return getStringFromWasm0(r0, r1);
        } finally {
            wasm.__wbindgen_add_to_stack_pointer(16);
            wasm.__wbindgen_free(r0, r1);
        }
    }
    /**
    * @param {number} index
    * @returns {string}
    */
    getSource(index) {
        try {
            const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
            wasm.sourcemap_getSource(retptr, this.ptr, index);
            var r0 = getInt32Memory0()[retptr / 4 + 0];
            var r1 = getInt32Memory0()[retptr / 4 + 1];
            return getStringFromWasm0(r0, r1);
        } finally {
            wasm.__wbindgen_add_to_stack_pointer(16);
            wasm.__wbindgen_free(r0, r1);
        }
    }
    /**
    * @param {string} name
    * @returns {number}
    */
    getNameIndex(name) {
        var ptr0 = passStringToWasm0(name, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        var ret = wasm.sourcemap_getNameIndex(this.ptr, ptr0, len0);
        return ret;
    }
    /**
    * @param {string} source
    * @returns {any}
    */
    getSourceIndex(source) {
        var ptr0 = passStringToWasm0(source, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        var ret = wasm.sourcemap_getSourceIndex(this.ptr, ptr0, len0);
        return takeObject(ret);
    }
    /**
    * @param {Int32Array} mappings_arr
    */
    addIndexedMappings(mappings_arr) {
        var ptr0 = passArray32ToWasm0(mappings_arr, wasm.__wbindgen_malloc);
        var len0 = WASM_VECTOR_LEN;
        wasm.sourcemap_addIndexedMappings(this.ptr, ptr0, len0);
    }
    /**
    * @returns {any}
    */
    toBuffer() {
        var ret = wasm.sourcemap_toBuffer(this.ptr);
        return takeObject(ret);
    }
    /**
    * @param {SourceMap} previous_map_instance
    * @param {number} line_offset
    * @returns {any}
    */
    addSourceMap(previous_map_instance, line_offset) {
        _assertClass(previous_map_instance, SourceMap);
        var ret = wasm.sourcemap_addSourceMap(this.ptr, previous_map_instance.ptr, line_offset);
        return takeObject(ret);
    }
    /**
    * @param {string} source
    * @param {string} source_content
    * @returns {any}
    */
    setSourceContentBySource(source, source_content) {
        var ptr0 = passStringToWasm0(source, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        var ptr1 = passStringToWasm0(source_content, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len1 = WASM_VECTOR_LEN;
        var ret = wasm.sourcemap_setSourceContentBySource(this.ptr, ptr0, len0, ptr1, len1);
        return takeObject(ret);
    }
    /**
    * @param {string} source
    * @returns {any}
    */
    getSourceContentBySource(source) {
        var ptr0 = passStringToWasm0(source, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        var ret = wasm.sourcemap_getSourceContentBySource(this.ptr, ptr0, len0);
        return takeObject(ret);
    }
    /**
    * @param {string} source
    * @param {string} source_content
    * @param {number} line_offset
    * @returns {any}
    */
    addEmptyMap(source, source_content, line_offset) {
        var ptr0 = passStringToWasm0(source, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        var ptr1 = passStringToWasm0(source_content, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len1 = WASM_VECTOR_LEN;
        var ret = wasm.sourcemap_addEmptyMap(this.ptr, ptr0, len0, ptr1, len1, line_offset);
        return takeObject(ret);
    }
    /**
    * @param {SourceMap} previous_map_instance
    * @returns {any}
    */
    extends(previous_map_instance) {
        _assertClass(previous_map_instance, SourceMap);
        var ret = wasm.sourcemap_extends(this.ptr, previous_map_instance.ptr);
        return takeObject(ret);
    }
    /**
    * @param {number} generated_line
    * @param {number} generated_column
    * @returns {any}
    */
    findClosestMapping(generated_line, generated_column) {
        var ret = wasm.sourcemap_findClosestMapping(this.ptr, generated_line, generated_column);
        return takeObject(ret);
    }
    /**
    * @param {number} generated_line
    * @param {number} generated_line_offset
    * @returns {any}
    */
    offsetLines(generated_line, generated_line_offset) {
        var ret = wasm.sourcemap_offsetLines(this.ptr, generated_line, generated_line_offset);
        return takeObject(ret);
    }
    /**
    * @param {number} generated_line
    * @param {number} generated_column
    * @param {number} generated_column_offset
    * @returns {any}
    */
    offsetColumns(generated_line, generated_column, generated_column_offset) {
        var ret = wasm.sourcemap_offsetColumns(this.ptr, generated_line, generated_column, generated_column_offset);
        return takeObject(ret);
    }
}

async function load(module, imports) {
    if (typeof Response === 'function' && module instanceof Response) {
        if (typeof WebAssembly.instantiateStreaming === 'function') {
            try {
                return await WebAssembly.instantiateStreaming(module, imports);

            } catch (e) {
                if (module.headers.get('Content-Type') != 'application/wasm') {
                    console.warn("`WebAssembly.instantiateStreaming` failed because your server does not serve wasm with `application/wasm` MIME type. Falling back to `WebAssembly.instantiate` which is slower. Original error:\n", e);

                } else {
                    throw e;
                }
            }
        }

        const bytes = await module.arrayBuffer();
        return await WebAssembly.instantiate(bytes, imports);

    } else {
        const instance = await WebAssembly.instantiate(module, imports);

        if (instance instanceof WebAssembly.Instance) {
            return { instance, module };

        } else {
            return instance;
        }
    }
}

async function init(input) {
    if (typeof input === 'undefined') {
        input = new URL('parcel_sourcemap_wasm_bg.wasm', import.meta.url);
    }
    const imports = {};
    imports.wbg = {};
    imports.wbg.__wbindgen_json_serialize = function(arg0, arg1) {
        const obj = getObject(arg1);
        var ret = JSON.stringify(obj === undefined ? null : obj);
        var ptr0 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        getInt32Memory0()[arg0 / 4 + 1] = len0;
        getInt32Memory0()[arg0 / 4 + 0] = ptr0;
    };
    imports.wbg.__wbg_new_3e1ee746fe308c9f = function(arg0, arg1) {
        var ret = new Error(getStringFromWasm0(arg0, arg1));
        return addHeapObject(ret);
    };
    imports.wbg.__wbindgen_is_undefined = function(arg0) {
        var ret = getObject(arg0) === undefined;
        return ret;
    };
    imports.wbg.__wbg_length_2d56cb37075fcfb1 = function(arg0) {
        var ret = getObject(arg0).length;
        return ret;
    };
    imports.wbg.__wbindgen_memory = function() {
        var ret = wasm.memory;
        return addHeapObject(ret);
    };
    imports.wbg.__wbg_buffer_9e184d6f785de5ed = function(arg0) {
        var ret = getObject(arg0).buffer;
        return addHeapObject(ret);
    };
    imports.wbg.__wbg_new_e8101319e4cf95fc = function(arg0) {
        var ret = new Uint8Array(getObject(arg0));
        return addHeapObject(ret);
    };
    imports.wbg.__wbindgen_object_drop_ref = function(arg0) {
        takeObject(arg0);
    };
    imports.wbg.__wbg_set_e8ae7b27314e8b98 = function(arg0, arg1, arg2) {
        getObject(arg0).set(getObject(arg1), arg2 >>> 0);
    };
    imports.wbg.__wbindgen_json_parse = function(arg0, arg1) {
        var ret = JSON.parse(getStringFromWasm0(arg0, arg1));
        return addHeapObject(ret);
    };
    imports.wbg.__wbindgen_number_new = function(arg0) {
        var ret = arg0;
        return addHeapObject(ret);
    };
    imports.wbg.__wbg_newwithbyteoffsetandlength_e57ad1f2ce812c03 = function(arg0, arg1, arg2) {
        var ret = new Uint8Array(getObject(arg0), arg1 >>> 0, arg2 >>> 0);
        return addHeapObject(ret);
    };
    imports.wbg.__wbindgen_string_new = function(arg0, arg1) {
        var ret = getStringFromWasm0(arg0, arg1);
        return addHeapObject(ret);
    };
    imports.wbg.__wbindgen_throw = function(arg0, arg1) {
        throw new Error(getStringFromWasm0(arg0, arg1));
    };
    imports.wbg.__wbindgen_rethrow = function(arg0) {
        throw takeObject(arg0);
    };

    if (typeof input === 'string' || (typeof Request === 'function' && input instanceof Request) || (typeof URL === 'function' && input instanceof URL)) {
        input = fetch(input);
    }



    const { instance, module } = await load(await input, imports);

    wasm = instance.exports;
    init.__wbindgen_wasm_module = module;

    return wasm;
}

export default init;
